src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/service/haflow/HaFlowReadService.java

```diff
@@ -0,0 +1,217 @@
+                FlowProtectedPathDto.FlowProtectedPathDtoBuilder protectedDtoBuilder =
+                        FlowProtectedPathDto.builder();
+                // At least 2 paths required to calculate Y-point.
+                if (isProperPaths(protectedForwardPaths.values())) {
+                    List<PathSegment> pathSegments =
+                            IntersectionComputer.calculatePathIntersectionFromSource(protectedForwardPaths.values());
+                    protectedDtoBuilder.forwardPath(
+                            FlowPathMapper.INSTANCE.mapToPathNodes(sharedEndpoint, pathSegments, null));
+                }
+                if (isProperPaths(protectedReversePaths.values())) {
+                    List<PathSegment> pathSegments =
+                            IntersectionComputer.calculatePathIntersectionFromDest(protectedReversePaths.values());
+                    protectedDtoBuilder.reversePath(
+                            FlowPathMapper.INSTANCE.mapToPathNodes(null, pathSegments, sharedEndpoint));
+                }
+                sharedPath.setProtectedPath(protectedDtoBuilder.build());
+            }
+
+            return new HaFlowPathsResponse(sharedPath, subFlowPathDtos, Collections.emptyMap());
+        });
+    }
+
+    private FlowPathDtoBuilder buildFlowPathDto(HaFlow haFlow, FlowPath forwardPath, FlowPath reversePath) {
+        return FlowPathDto.builder()
+                .id(forwardPath.getHaSubFlowId())
+                .forwardPath(FlowPathMapper.INSTANCE.mapToPathNodes(forwardPath, getInPort(haFlow, forwardPath),
+                        getOutPort(haFlow, forwardPath)))
+                .reversePath(FlowPathMapper.INSTANCE.mapToPathNodes(reversePath, getInPort(haFlow, reversePath),
+                        getOutPort(haFlow, reversePath)));
+    }
+
+    private FlowProtectedPathDto buildProtectedPathDto(HaFlow haFlow, FlowPath forwardPath, FlowPath reversePath) {
+        FlowProtectedPathDtoBuilder builder = FlowProtectedPathDto.builder();
+        if (forwardPath != null) {
+            builder.forwardPath(FlowPathMapper.INSTANCE.mapToPathNodes(forwardPath, getInPort(haFlow, forwardPath),
+                    getOutPort(haFlow, forwardPath)));
+        }
+        if (reversePath != null) {
+            builder.reversePath(FlowPathMapper.INSTANCE.mapToPathNodes(reversePath, getInPort(haFlow, reversePath),
+                    getOutPort(haFlow, reversePath)));
+        }
+        return builder.build();
+    }
+
+    private int getInPort(HaFlow haFlow, FlowPath flowPath) {
+        return FlowSideAdapter.makeIngressAdapter(haFlow, flowPath).getEndpoint().getPortNumber();
+    }
+
+    private int getOutPort(HaFlow haFlow, FlowPath flowPath) {
+        return FlowSideAdapter.makeEgressAdapter(haFlow, flowPath).getEndpoint().getPortNumber();
+    }
+
+    private boolean isProperPaths(Collection<FlowPath> paths) {
```

Comment:
I'm not sure that it is a good name. One switch path is also proper path

Reply:
As we are checking 2 conditions, not only one switches paths, I meant to name the method like: "Are these paths good enough to be used inside the 'IntersectionComputer.calculatePathIntersectionFromSource' method". Or the shorter version which you see. If you have a better idea feel free to share it and I'll update the name of the method.

Reply:
maybe `arePathsIntersectible`

Reply:
Thanks for the idea. I would go with the same by mining, but a little bit simple in writing name `arePathsCrossed`

Reply:
Fixed if agree with the name